File: tictactoe.pl
% win conditions

win(brd, plyr) :- hor(brd, plyr).
win(brd, plyr) :- ver(brd, plyr).
win(brd, plyr) :- diag(brd, plyr).

hor(brd, plyr) :- brd = [plyr,plyr,plyr,_,_,_,_,_,_].
hor(brd, plyr) :- brd = [_,_,_,plyr,plyr,plyr,_,_,_].
hor(brd, plyr) :- brd = [_,_,_,_,_,_,plyr,plyr,plyr].

ver(brd, plyr) :- brd = [plyr,_,_,plyr,_,_,plyr,_,_].
ver(brd, plyr) :- brd = [_,plyr,_,_,plyr,_,_,plyr,_].
ver(brd, plyr) :- brd = [_,_,plyr,_,_,plyr,_,_,plyr].

diag(brd, plyr) :- brd = [plyr,_,_,_,plyr,_,_,_,plyr].
diag(brd, plyr) :- brd = [_,_,plyr,_,plyr,_,plyr,_,_].

other(x,o).
other(o,x).

% game initialization

game(brd, plyr) :- 
  win(brd, plyr),
  !,
  write([plyr, plyr, wins]).
game(brd, plyr) :- 
  other(plyr,otherplyr),
  move(brd,plyr,newbrd),
  !,
  print(newbrd),
  game(newbrd,otherplyr).

% move procdure

move([-,B,C,D,E,F,G,H,I], plyr, [plyr,B,C,D,E,F,G,H,I]).
move([A,-,C,D,E,F,G,H,I], plyr, [A,plyr,C,D,E,F,G,H,I]).
move([A,B,-,D,E,F,G,H,I], plyr, [A,B,plyr,D,E,F,G,H,I]).
move([A,B,C,-,E,F,G,H,I], plyr, [A,B,C,plyr,E,F,G,H,I]).
move([A,B,C,D,-,F,G,H,I], plyr, [A,B,C,D,plyr,F,G,H,I]).
move([A,B,C,D,E,-,G,H,I], plyr, [A,B,C,D,E,plyr,G,H,I]).
move([A,B,C,D,E,F,-,H,I], plyr, [A,B,C,D,E,F,plyr,H,I]).
move([A,B,C,D,E,F,G,-,I], plyr, [A,B,C,D,E,F,G,plyr,I]).
move([A,B,C,D,E,F,G,H,-], plyr, [A,B,C,D,E,F,G,H,plyr]).

print([A,B,C,D,E,F,G,H,I]) :- 
  write([A,B,C]), nl,
  write([D,E,F]), nl,
  write([G,H,I]), nl,
  nl.

x_can_win(brd) :- move(brd, x, newbrd), win(newbrd, x).

% respond: computers reponse

respond(brd,newbrd) :- 
  move(brd, o, newbrd),
  win(newbrd, o),
  !.
respond(brd,newbrd) :-
  move(brd, o, newbrd), 
  not(x_can_win(newbrd)).
respond(brd,newbrd) :-
  move(brd, o, newbrd).
respond(brd,newbrd) :-
  not(member(-,brd)),
  !, 
  write('Tie!'), nl,
  newbrd = brd.

% player turn

xmove([-,B,C,D,E,F,G,H,I], 1, [x,B,C,D,E,F,G,H,I]).
xmove([A,-,C,D,E,F,G,H,I], 2, [A,x,C,D,E,F,G,H,I]).
xmove([A,B,-,D,E,F,G,H,I], 3, [A,B,x,D,E,F,G,H,I]).
xmove([A,B,C,-,E,F,G,H,I], 4, [A,B,C,x,E,F,G,H,I]).
xmove([A,B,C,D,-,F,G,H,I], 5, [A,B,C,D,x,F,G,H,I]).
xmove([A,B,C,D,E,-,G,H,I], 6, [A,B,C,D,E,x,G,H,I]).
xmove([A,B,C,D,E,F,-,H,I], 7, [A,B,C,D,E,F,x,H,I]).
xmove([A,B,C,D,E,F,G,-,I], 8, [A,B,C,D,E,F,G,x,I]).
xmove([A,B,C,D,E,F,G,H,-], 9, [A,B,C,D,E,F,G,H,x]).
xmove(brd, _, brd) :- write('Invalid move.'), nl.

% game start

play :- howto, playon([-,-,-,-,-,-,-,-,-]).


% instructions on how to play the game

howto :-
  write('You play X.'),
  nl,
  write('Enter Integer where you want to place X: '),
  nl,
  print([1,2,3,4,5,6,7,8,9]).

% recursive procedure to play the game

playon(brd) :- win(brd, x), write('Player wins the game!').
playon(brd) :- win(brd, o), write('Computer wins the game!').
playon(brd) :- read(N),
  xmove(brd, N, newbrd), 
  print(newbrd),
  respond(newbrd, newnewbrd), 
  print(newnewbrd),
  playon(newnewbrd).

File: tic-tac-toe.py
# TicTacToe using minimax

# Game Board
def Gameboard(board):
    print()
    for i in range(0, 9):
        if((i > 0) and ((i % 3) == 0)):
            print()
        if(board[i] == 0):
            print("---", end=" ")
        if (board[i] == 1):
            print(" X ", end=" ")
        if(board[i] == -1):
            print(" O ", end=" ")
    print()
    print()

# Player X
def x_move(board):
    loc = int(input("Enter a number between 1-9 to place X : "))
    while(board[loc-1] != 0):
        print("Invalid Move! Try again!")
        loc = int(input("Enter a number between 1-9 to place X : "))
    board[loc-1] = 1

# Player O
def o_move(board):
    loc = int(input("Enter a number between 1-9 to place O : "))
    while(board[loc-1] != 0):
        print("Invalid Move! Try again!")
        loc = int(input("Enter a number between 1-9 to place O : "))
    board[loc-1] = -1

# Win conditions
def condition(board):
    win = [[0, 1, 2], [3, 4, 5], [6, 7, 8],
               [0, 3, 6], [1, 4, 7], [2, 5, 8],
               [0, 4, 8], [2, 4, 6]]

    for i in range(0, 8):
        if(board[win[i][0]] != 0 and
           board[win[i][0]] == board[win[i][1]] and
           board[win[i][0]] == board[win[i][2]]):
            return board[win[i][2]]
    return 0

# Minimax


def minimax(board, player):
    a = condition(board)
    if(a != 0):
        return (a*player)
    loc = -1
    value = -2
    for i in range(0, 9):
        if(board[i] == 0):
            board[i] = player
            score = -minimax(board, (player*-1))
            if(score > value):
                value = score
                loc = i
            board[i] = 0

    if(loc == -1):
        return 0
    return value

# Computer Turn using Minimax
def comp(board):
    loc = -1
    value = -2
    for i in range(0, 9):
        if(board[i] == 0):
            board[i] = 1
            score = -minimax(board, -1)
            board[i] = 0
            if(score > value):
                value = score
                loc = i

    board[loc] = 1

# Player vs. Computer mode
def pvc(board):
    print()
    print("Computer : X Vs. You : O")
    player = int(input("Enter to play 1st or 2nd: "))
    for i in range(0, 9):
        if(condition(board) != 0):
            break
        if((i+player) % 2 == 0):
            comp(board)
        else:
            Gameboard(board)
            o_move(board)

# Player vs. Player mode
def pvp(board):
    for i in range(0, 9):
        if(condition(board) != 0):
            break
        if((i) % 2 == 0):
            Gameboard(board)
            x_move(board)
        else:
            Gameboard(board)
            o_move(board)

def winner(board):
    x = condition(board)
    if(x == 0):
        Gameboard(board)
        print("The Game is a Draw!")
        print()
    if(x == 1):
        Gameboard(board)
        print("Player X wins the Game!")
        print()
    if(x == -1):
        Gameboard(board)
        print("Player O wins the Game!")
        print()


# Driver Code
print("====================")
print("TIC-TAC-TOE using AI")
print("====================")

board = [
            0, 0, 0,
            0, 0, 0,
            0, 0, 0,
        ]

print("\n1. Player vs. Computer \n2. Player vs. Player")
op = int(input("Select the Game Mode: "))
if(op == 1):
    pvc(board)
else:
    pvp(board)

winner(board)
